### PR TITLE
NEW Can filter on extrafields date on lists

### PR DESCRIPTION
**Pull request by atm-florianm**
# FIX
## Problem description
Currently, lists are filterable on virtually every kind of extrafield except dates (they are sortable, but no input is shown for filtering them). Furthermore, if an input were provided, the user would have to enter the date corresponding to the exact timestamp of the desired extrafields.

## This PR
This PR adds a date range filter (From date X to date Y) to date extrafield columns on list views.

Note: as  'date' and 'datetime' extrafields are handled together, I didn't add a specific handler for 'datetime', so 'datetime' extrafields get filtered with a simple date range (not a more precise date+time range).

### CODE REVIEW

File: htdocs/core/class/extrafields.class.php [hunk 1]
```diff
@@ -2162,9 +2176,28 @@ public function getOptionalsFromPost($extrafieldsobjectkey, $keyprefix = '', $ke
...
 				{
-					if (!GETPOSTISSET($keysuffix."options_".$key.$keyprefix."year")) continue; // Value was not provided, we should not set it.
-					// Clean parameters
-					$value_key = dol_mktime(GETPOST($keysuffix."options_".$key.$keyprefix."hour", 'int'), GETPOST($keysuffix."options_".$key.$keyprefix."min", 'int'), 0, GETPOST($keysuffix."options_".$key.$keyprefix."month", 'int'), GETPOST($keysuffix."options_".$key.$keyprefix."day", 'int'), GETPOST($keysuffix."options_".$key.$keyprefix."year", 'int'));
+					$dateparamname_start = $keysuffix . 'options_' . $key . $keyprefix . '_start';
+					$dateparamname_end   = $keysuffix . 'options_' . $key . $keyprefix . '_end';
+					if (GETPOSTISSET($dateparamname_start . 'year') && GETPOSTISSET($dateparamname_end . 'year')) {
+						// values provided as a date pair (start date + end date), each date being broken down as year, month, day, etc.
+						$value_key = array(
+							'start' => mktime(
```
Comment (FHenry) [inline htdocs/core/class/extrafields.class.php:2184]: You mktime and not dol_mktime ?

Comment (atm-florianm) [reply htdocs/core/class/extrafields.class.php]: Thanks, I’ll change that.

File: htdocs/core/tpl/extrafields_list_search_param.tpl.php [hunk 2]
```diff
@@ -16,6 +16,21 @@
     {
         $crit = $val;
         $tmpkey = preg_replace('/'.$search_options_pattern.'/', '', $key);
+		if (is_array($val) && array_key_exists('start', $val) && array_key_exists('end', $val)) {
+			// date range from list filters is stored as array('start' => <timestamp>, 'end' => <timestamp>)
+			// start date
+			$param .= '&' . $search_options_pattern.$tmpkey.'_startyear='  . intval(dol_print_date($val['start'], '%Y'));
```
Comment (eldy) [inline htdocs/core/tpl/extrafields_list_search_param.tpl.php:22]: why the "intval" ?  The dol_print_date should return a string (what we need for the concat)

Comment (atm-florianm) [reply htdocs/core/tpl/extrafields_list_search_param.tpl.php]: I removed the intval() as it didn't really make sense, as you pointed out.

**Pull request merged**